ucp/benchmarks/backends/tornado.py

```diff
@@ -0,0 +1,121 @@
+import asyncio
+from time import monotonic
+
+from tornado.iostream import StreamClosedError
+from tornado.tcpclient import TCPClient
+from tornado.tcpserver import TCPServer
+
+from ucp.benchmarks.backends.base import BaseClient, BaseServer
+
+
+class TornadoServer(BaseServer):
+    def __init__(self, args, xp, queue):
+        self.args = args
+        self.xp = xp
+        self.queue = queue
+
+    def has_cuda_support() -> bool:
+        return False
+
+    def _start_listener(self, server, port):
+        if port is not None:
+            server.listen(port)
+        else:
+            for i in range(10000, 60000):
+                try:
+                    server.listen(i)
+                except OSError:
+                    continue
+                else:
+                    port = i
+                    break
+
+        return port
+
+    async def run(self):
+        args = self.args
+        xp = self.xp
+
+        global event
+        event = asyncio.Event()
+
+        class TransferServer(TCPServer):
+            async def handle_stream(self, stream, address):
+                global event
+                msg_recv_list = []
+                if not args.reuse_alloc:
+                    for _ in range(args.n_iter + args.n_warmup_iter):
+                        msg_recv_list.append(xp.zeros(args.n_bytes, dtype="u1"))
+                else:
+                    t = xp.zeros(args.n_bytes, dtype="u1")
+                    for _ in range(args.n_iter + args.n_warmup_iter):
+                        msg_recv_list.append(t)
+
+                assert msg_recv_list[0].nbytes == args.n_bytes
+
+                for i in range(args.n_iter + args.n_warmup_iter):
+                    try:
+                        await stream.read_into(msg_recv_list[i].data)
+                        await stream.write(msg_recv_list[i].data)
```

Comment:
I know this mimics the other benchmarks but in the "alloc each time case" I would have though a more realistic setup would be to allocate as we go (rather than allocate niter buffers up front). WDYT?

Reply:
Hmm, I wouldn't mind, but it seems using `read_bytes` instead of `read_into` makes it much slower, see below.

<details><summary>`read_into` (current)</summary>

```
$ python -m ucp.benchmarks.send_recv -l tornado
Server Running at 10.33.225.163:10000
Client connecting to server at 10.33.225.163:10000
Roundtrip benchmark
================================================================================
Iterations                | 10
Bytes                     | 9.54 MiB
Object type               | numpy
Reuse allocation          | False
================================================================================
Device(s)                 | CPU-only
Server CPU                | affinity not set
Client CPU                | affinity not set
================================================================================
Bandwidth (average)       | 2.10 GiB/s
Bandwidth (median)        | 2.10 GiB/s
Latency (average)         | 4427092 ns
Latency (median)          | 4433597 ns
================================================================================
Iterations                | Bandwidth, Latency
--------------------------------------------------------------------------------
0                         | 2.10 GiB/s, 4432572ns
1                         | 2.09 GiB/s, 4446343ns
2                         | 2.09 GiB/s, 4447852ns
3                         | 2.10 GiB/s, 4434623ns
4                         | 2.11 GiB/s, 4414575ns
5                         | 2.11 GiB/s, 4419028ns
6                         | 2.11 GiB/s, 4418157ns
7                         | 2.08 GiB/s, 4472294ns
8                         | 2.10 GiB/s, 4435407ns
9                         | 2.14 GiB/s, 4350074ns
```

</details>

<details><summary>`read_bytes` (proposed)</summary>

```
$ python -m ucp.benchmarks.send_recv -l tornado
Server Running at 10.33.225.163:10000
Client connecting to server at 10.33.225.163:10000
Roundtrip benchmark
================================================================================
Iterations                | 10
Bytes                     | 9.54 MiB
Object type               | numpy
Reuse allocation          | False
================================================================================
Device(s)                 | CPU-only
Server CPU                | affinity not set
Client CPU                | affinity not set
================================================================================
Bandwidth (average)       | 1.27 GiB/s
Bandwidth (median)        | 1.27 GiB/s
Latency (average)         | 7312802 ns
Latency (median)          | 7327963 ns
================================================================================
Iterations                | Bandwidth, Latency
--------------------------------------------------------------------------------
0                         | 1.27 GiB/s, 7342915ns
1                         | 1.27 GiB/s, 7331363ns
2                         | 1.27 GiB/s, 7361351ns
3                         | 1.27 GiB/s, 7324030ns
4                         | 1.27 GiB/s, 7324563ns
5                         | 1.26 GiB/s, 7362499ns
6                         | 1.27 GiB/s, 7321720ns
7                         | 1.27 GiB/s, 7348971ns
8                         | 1.27 GiB/s, 7314609ns
9                         | 1.31 GiB/s, 7096002ns
```

</details>

To avoid giving UCX-Py a potentially unfair advantage, it seems better to keep it as is. Also, not committing it yet, but below is diff for the changes I've made, in case you can spot something I did wrong and created an unfair advantage for us:

<details><summary>`read_bytes` patch</summary>

```diff
diff --git a/ucp/benchmarks/backends/tornado.py b/ucp/benchmarks/backends/tornado.py
index 3076d99..8bd90f1 100644
--- a/ucp/benchmarks/backends/tornado.py
+++ b/ucp/benchmarks/backends/tornado.py
@@ -42,20 +42,21 @@ class TornadoServer(BaseServer):
             async def handle_stream(self, stream, address):
                 global event
                 msg_recv_list = []
-                if not args.reuse_alloc:
-                    for _ in range(args.n_iter + args.n_warmup_iter):
-                        msg_recv_list.append(xp.zeros(args.n_bytes, dtype="u1"))
-                else:
+                if args.reuse_alloc:
                     t = xp.zeros(args.n_bytes, dtype="u1")
                     for _ in range(args.n_iter + args.n_warmup_iter):
                         msg_recv_list.append(t)

-                assert msg_recv_list[0].nbytes == args.n_bytes
+                    assert msg_recv_list[0].nbytes == args.n_bytes

                 for i in range(args.n_iter + args.n_warmup_iter):
                     try:
-                        await stream.read_into(msg_recv_list[i].data)
-                        await stream.write(msg_recv_list[i].data)
+                        if args.reuse_alloc:
+                            await stream.read_into(msg_recv_list[i].data)
+                            await stream.write(msg_recv_list[i].data)
+                        else:
+                            msg_recv_list.append(await stream.read_bytes(args.n_bytes))
+                            await stream.write(msg_recv_list[i])
                     except StreamClosedError as e:
                         print(e)
                         break
@@ -89,18 +90,18 @@ class TornadoClient(BaseClient):

         msg_send_list = []
         msg_recv_list = []
-        if not self.args.reuse_alloc:
-            for i in range(self.args.n_iter + self.args.n_warmup_iter):
-                msg_send_list.append(self.xp.arange(self.args.n_bytes, dtype="u1"))
-                msg_recv_list.append(self.xp.zeros(self.args.n_bytes, dtype="u1"))
-        else:
+        if self.args.reuse_alloc:
             t1 = self.xp.arange(self.args.n_bytes, dtype="u1")
             t2 = self.xp.zeros(self.args.n_bytes, dtype="u1")
             for i in range(self.args.n_iter + self.args.n_warmup_iter):
                 msg_send_list.append(t1)
                 msg_recv_list.append(t2)
+            assert msg_recv_list[0].nbytes == self.args.n_bytes
+        else:
+            for i in range(self.args.n_iter + self.args.n_warmup_iter):
+                msg_send_list.append(self.xp.arange(self.args.n_bytes, dtype="u1"))
+
         assert msg_send_list[0].nbytes == self.args.n_bytes
-        assert msg_recv_list[0].nbytes == self.args.n_bytes

         if self.args.cuda_profile:
             self.xp.cuda.profiler.start()
@@ -109,7 +110,10 @@ class TornadoClient(BaseClient):
             start = monotonic()

             await stream.write(msg_send_list[i].data)
-            await stream.read_into(msg_recv_list[i].data)
+            if self.args.reuse_alloc:
+                await stream.read_into(msg_recv_list[i].data)
+            else:
+                msg_recv_list.append(await stream.read_bytes(self.args.n_bytes))

             stop = monotonic()
             if i >= self.args.n_warmup_iter:
```

</details>

Reply:
I think I meant something like:
```
if self.args.reuse_alloc:
   await stream.read_into(msg_recv_list[i].data)
else:
   buf = self.xp.empty(self.args.n_bytes, dtype="u1")
   await stream.read_into(buf.data)
   msg_recv_list.append(buf)
```

Reply:
I see your point, I agree this makes more sense, we are not necessarily timing the allocation this way. If you don't mind I'll do it in a follow-up PR, then I'll do a similar change to the UCX-Py benchmarks together with this, that will ensure that all of the backends are timing things in an equivalent manner.

Reply:
Opened https://github.com/rapidsai/ucx-py/pull/880 to cover this, but this needs to go in first.